### PR TITLE
Split LinkageMonitor.run with test

### LINKED ISSUE #742: LinkageMonitor.run too complex to test
LinkageMonitor.run too complex to test

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #742 

In draft. I'll add tests for new functions but welcome feedback if any.

### CODE REVIEW

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java [hunk 1]
```diff
@@ -73,15 +73,9 @@ private void run(String groupId, String artifactId)
...
-    ImmutableList<String> snapshotCoordinates =
-        snapshot.getManagedDependencies().stream()
-            .map(Artifacts::toCoordinates)
-            .collect(toImmutableList());
+    // Comparing coordinates because DefaultArtifact does not override equals
+    ImmutableList<String> baselineCoordinates = coordinatesList(snapshot.getManagedDependencies());
+    ImmutableList<String> snapshotCoordinates = coordinatesList(snapshot.getManagedDependencies());
     if (baselineCoordinates.equals(snapshotCoordinates)) {
       System.out.println(
           "The content of the snapshot BOM and the original BOM are the same. Not running"
```
Comment (elharo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: it's head, not snapshot

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java [hunk 2]
```diff
@@ -101,32 +95,12 @@ private void run(String groupId, String artifactId)
...
-      StringBuilder message =
-          new StringBuilder(
-              "The following problem"
-                  + (problemSize > 1 ? "s" : "")
-                  + " in the baseline no longer appear in the snapshot:\n");
-      for (SymbolProblem problem : fixedProblems) {
-        message.append(problem + "\n");
-      }
-      System.out.println(message.toString());
+      System.out.println(formatMessageForFixedError(fixedProblems));
```
Comment (elharo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: I suspect System.out.println is not the right place to putput, but we can think about that later. Maybe a logger? I'm not sure.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java [hunk 3]
```diff
@@ -135,6 +109,49 @@ private void run(String groupId, String artifactId)
     }
   }
 
+  private static ImmutableList<String> coordinatesList(List<Artifact> artifacts) {
```
Comment (elharo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: nice improvement

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java [hunk 4]
```diff
@@ -135,6 +109,49 @@ private void run(String groupId, String artifactId)
...
+    return artifacts.stream().map(Artifacts::toCoordinates).collect(toImmutableList());
+  }
+
+  /**
+   * Returns an error message if {@code snapshotSymbolProblems} contains new problems compared with
```
Comment (elharo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: This is an odd pattern. Perhaps we should split the detection of the error from the formatting of the message. Or perhaps the comment isn' correct? I don't see this method returning null.

Comment (suztomo) [reply linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: The javadoc was outdated. Initially I was thinking to mix error detection and formatting message but it's already separated.

**Comment by suztomo** [comment]
@elharo PTAL.

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java [hunk 5]
```diff
@@ -135,6 +109,49 @@ private void run(String groupId, String artifactId)
...
+  private static ImmutableList<String> coordinatesList(List<Artifact> artifacts) {
+    return artifacts.stream().map(Artifacts::toCoordinates).collect(toImmutableList());
+  }
+
+  /**
+   * Returns a message on {@code snapshotSymbolProblems} that do not exist in {@code
+   * baselineProblems}.
+   */
+  @VisibleForTesting
+  static String formatMessageForNewError(
```
Comment (elharo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: perhaps just newErrorsMessage(); your choice

File: linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java [hunk 6]
```diff
@@ -135,6 +109,49 @@ private void run(String groupId, String artifactId)
...
+                "  referenced from %s (%s)\n",
+                classFile.getClassName(), classFile.getJar().getFileName()));
+      }
+    }
+    return message.toString();
+  }
+
+  /** Returns a message on {@code fixedProblems}. */
+  @VisibleForTesting
+  static String formatMessageForFixedError(Set<SymbolProblem> fixedProblems) {
```
Comment (elharo) [inline linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java]: fixedErrorsMessage()?

**Pull request merged**